source/common/grpc/google_grpc_creds_impl.h

```diff
@@ -0,0 +1,14 @@
+#pragma once
+
+#include "envoy/api/v2/core/grpc_service.pb.h"
+
+#include "grpc++/grpc++.h"
+
+namespace Envoy {
+namespace Grpc {
+
+std::shared_ptr<grpc::ChannelCredentials>
+getGoogleGrpcChannelCredentials(const envoy::api::v2::core::GrpcService& grpc_service);
+
+} // namespace Grpc
+} // namespace Envoy
```

Comment:
We'll definitely needs tests. You should write a dummy resolver that works async so that we can simulate what happens if we have network delays etc.

Reply:
I have added a test with a static access_token.  I've done local testing with delays to show that the main thread is not blocked, however I'm still a bit unsure how to add a test for this into the test framework.